qiskit/algorithms/linear_solvers/observables/matrix_functional.py

```diff
@@ -169,6 +169,7 @@ def evaluate_classically(self, solution: Union[np.array, QuantumCircuit]) -> flo
         # Check if it is QuantumCircuits and get the array from them
         if isinstance(solution, QuantumCircuit):
             solution = Statevector(solution).data
+        assert isinstance(solution, np.ndarray)
```

Comment:
This run-time assertion shouldn't be in code.  The "correct" way of doing this is `typing.cast`, but personally I don't think that should ever be present in run-time code either.  Mypy is probably failing to infer the type of `solution` after the line because `Statevector.data` isn't typed - the best solution is probably to get the typing there instead.